trace_decoder/src/typed_mpt.rs

```diff
@@ -0,0 +1,284 @@
+
+impl TriePath {
+    pub fn new(components: impl IntoIterator<Item = U4>) -> anyhow::Result<Self> {
+        Ok(TriePath(CopyVec::try_from_iter(components)?))
+    }
+    pub fn into_hash_left_padded(mut self) -> H256 {
+        for _ in 0..self.0.spare_capacity_mut().len() {
+            self.0.insert(0, U4::Dec00)
+        }
+        let mut packed = [0u8; 32];
+        AsNibbles(&mut packed).pack_from_slice(&self.0);
+        H256::from_slice(&packed)
+    }
+    fn from_address(address: Address) -> Self {
+        Self::from_hash(keccak_hash::keccak(address))
+    }
+    fn from_hash(H256(bytes): H256) -> Self {
+        Self::new(AsNibbles(bytes)).expect("32 bytes is 64 nibbles, which fits")
+    }
+    fn from_txn_ix(txn_ix: usize) -> Self {
+        TriePath::new(AsNibbles(rlp::encode(&txn_ix))).expect(
+            "\
+            rlp of an usize goes through a u64, which is 8 bytes,
+            which will be 9 bytes RLP'ed.
+            9 < 32
+        ",
+        )
+    }
+    fn into_nibbles(self) -> mpt_trie::nibbles::Nibbles {
+        let mut theirs = mpt_trie::nibbles::Nibbles::default();
+        for component in self.0 {
+            theirs.push_nibble_back(component as u8)
+        }
+        theirs
+    }
+    pub fn from_nibbles(mut theirs: mpt_trie::nibbles::Nibbles) -> Self {
+        let mut ours = CopyVec::new();
+        while !theirs.is_empty() {
+            ours.try_push(
+                U4::new(theirs.pop_next_nibble_front())
+                    .expect("mpt_trie returned an invalid nibble"),
+            )
+            .expect("mpt_trie should not have more than 64 nibbles")
+        }
+        Self(ours)
+    }
+}
+
+/// Map where keys are [up to 64 nibbles](TriePath), and values are either an
+/// out-of-band [hash](H256) or an inline [`rlp::Encodable`]/[`rlp::Decodable`]
+/// value.
+///
+/// [Merkle Patricia Trees](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie)
+/// are _maps_, where keys are typically _sequences_ of an _alphabet_.
+///
+/// Map values are typically indirect (i.e a _hash_),
+/// but in this structure may be stored _inline_.
+///
+/// Semantically equivalent to a [`mpt_trie::partial_trie::HashedPartialTrie`].
+#[derive(Debug, Clone)]
+struct TypedMpt<T> {
+    map: BTreeMap<TriePath, Either<H256, T>>,
+}
+
+impl<T> Default for TypedMpt<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> TypedMpt<T> {
+    pub const fn new() -> Self {
+        Self {
+            map: BTreeMap::new(),
+        }
+    }
+    // pub fn remove(&mut self, path: TriePath) -> Option<Either<H256, T>> {
+    //     self.map.remove(&path)
+    // }
+    pub fn insert(&mut self, path: TriePath, value: T) -> Option<Either<H256, T>> {
+        self.map.insert(path, Either::Right(value))
+    }
+    pub fn insert_branch(&mut self, path: TriePath, hash: H256) -> Option<Either<H256, T>> {
+        self.map.insert(path, Either::Left(hash))
+    }
```

Comment:
So I haven't been too involved in reviewing your major refactor PR, so apologies if I'm missing some context here, but I'm a bit surprised that the caller needs to "know" about nodes that are not leaves. Like more specifically, I mean given just the leaf nodes for a MPT/SMT trie, we should be able to deterministically reconstruct the exact same intermediate nodes just from only knowing the leaf nodes.

Reply:
> given just the leaf nodes for a MPT/SMT trie, we should be able to deterministically reconstruct the exact same intermediate nodes just from only knowing the leaf nodes.

That makes sense to me! But the witness format will be pruned - i.e it will contain hashes instead of full leaves:
https://github.com/0xPolygonZero/zk_evm/blob/7807a7db38f2ba88d48fdd800a34dc715752d337/trace_decoder/src/type1.rs#L132
https://github.com/0xPolygonZero/zk_evm/blob/7807a7db38f2ba88d48fdd800a34dc715752d337/trace_decoder/src/type1.rs#L57

So we need this API during tree construction, and during tree pruning (whenever I get to adding that).

Please let me know if I'm missing anything from my picture here!

Reply:
So I could be missing something as well, but even with pruning, I'm not sure if the user needs to manually insert `branch` nodes. Like given a sequence of `leaf`/`hash` node inserts, there should be one specific intermediate node structure that will exist according to the Ethereum spec. If we're allowing the library user to manually insert branch nodes, then I think this guarantee now relies on the user knowing what they are doing (I think...).

For pruning, I think all that we need to do is given a set of accesses (ie. keys), mark all nodes that are "hit" from traversing down the trie with each key. All nodes that are marked are not hashed out, and for the sections of the trie that are not hit, we hash out whatever parent is the highest up the trie.

Again, I could still be missing something here, but I don't think we want to expose inserting intermediate nodes to the API, except maybe for testing reasons.

Reply:
Right, I see the confusion. `insert_branch` does not insert a `Node::Branch`, it's semantically a `Node::Hash`. I've taken the terminology from [alloy_trie](https://docs.rs/alloy-trie/latest/alloy_trie/hash_builder/struct.HashBuilder.html#method.add_branch).

I'll change the wording as appropriate.

Reply:
Ahh makes sense, thanks!